File: assiist_back_end/db/repositories/interfaces/user_metrics_repository.py
from abc import ABC, abstractmethod
from typing import Dict, Optional

from assiist_back_end.models.user_metrics import UserMetrics

class UserMetricsRepository(ABC):
    """Abstract interface for user metrics data operations."""

    @abstractmethod
    async def get_metrics_for_contact(self, user_id: str, contact_id: str) -> Optional[UserMetrics]:
        """Gets metrics for a specific contact, creating initial metrics if they don't exist."""
        pass

    @abstractmethod
    async def increment_messages_sent(self, user_id: str, contact_id: str, idempotency_key: str) -> None:
        """Increments the messages sent counter for a contact, idempotently using the provided key."""
        pass

    @abstractmethod
    async def increment_notes_logged(self, user_id: str, contact_id: str) -> None:
        """Increments the notes logged counter for a contact. (Not idempotent, unlike messages_sent)"""
        pass

    @abstractmethod
    async def get_metrics_for_user(self, user_id: str) -> Dict[str, int]:
        """Gets total metrics across all contacts for a user."""
        pass 

File: assiist_front_end/lib/core/repositories/user_metrics_repository.dart
import 'package:assiist_front_end/core/models/user_metrics.dart';

abstract class UserMetricsRepository {
  Future<UserMetrics?> getMetricsForContact(String contactId);
  Future<void> incrementMessagesSent(String contactId);
  Future<void> incrementNotesLogged(String contactId);
  Future<Map<String, int>> getMetricsForUser();
}
